Rework STM32 ADC sequencer

gautierg-st opened the pull request:
The ADC sequencer was not supported on all STM32 series.

This PR adds the support of the sequencer for all STM32 series with the exception of STM32F1 that needs DMA for that (and which is not fully supported yet).

Overlays of tests/drivers/adc_api have been updated in order to check the sequencer in automatic daily tests.

Fixes #61100

gautierg-st committed:
drivers: adc: rework stm32 adc sequencer

Add sequencer support for all STM32 series.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for nucleo g071rb

Add multiple channels in overlay to test the sequencer.
Change clock source to correctly pass the test.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: stm32: remove channel count limit

Remove CHANNEL_COUNT limit used to check the channel bitmask.
This value was not applicable on STM32L1 where channel can go up to 31.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for nucleo f401re

Add multiple channels in overlay to test the sequencer.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for nucleo f091rc

Add multiple channels in overlay to test the sequencer.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for nucleo f207zg

Add multiple channels in overlay to test the sequencer.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for nucleo f429zi

Add multiple channels in overlay to test the sequencer.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for nucleo f746zg

Add multiple channels in overlay to test the sequencer.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for nucleo g474re

Add multiple channels in overlay to test the sequencer.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for nucleo l073rz

Add multiple channels in overlay to test the sequencer.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for nucleo l152re

Add multiple channels in overlay to test the sequencer.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for nucleo l552ze_q

Add multiple channels in overlay to test the sequencer.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for nucleo wb55rg

Add multiple channels in overlay to test the sequencer.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for nucleo wba52cg

Add multiple channels in overlay to test the sequencer.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for nucleo wl55jc

Add multiple channels in overlay to test the sequencer.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for stm32f3 disco

Add multiple channels in overlay to test the sequencer.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for stm32h573i dk

Add multiple channels in overlay to test the sequencer.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for stm32l562e dk

Add multiple channels in overlay to test the sequencer.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for disco l475 iot1

Add multiple channels in overlay to test the sequencer.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for b_u585i_iot02a

Add multiple channels in overlay to test the sequencer for ADC1.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: adc_api: boards: add sequence for b_u585i_iot02a

Add multiple channels in overlay to test the sequencer for ADC4.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -72,12 +72,43 @@ LOG_MODULE_REGISTER(adc_stm32);
+	defined(CONFIG_SOC_SERIES_STM32F0X) || \
+	defined(CONFIG_SOC_SERIES_STM32G0X) || \
+	defined(CONFIG_SOC_SERIES_STM32L0X) || \
+	defined(CONFIG_SOC_SERIES_STM32U5X) || \
```
Is it expected that U5 is both confirgurable and not configurable ?

gautierg-st replied:
It is, I even put a comment about it below. ON U5, ADC1 and 2 are fully configurable, but not ADC4

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -72,12 +72,43 @@ LOG_MODULE_REGISTER(adc_stm32);
-	!defined(CONFIG_SOC_SERIES_STM32L0X) && \
-	!defined(CONFIG_SOC_SERIES_STM32WBAX) && \
-	!defined(CONFIG_SOC_SERIES_STM32WLX)
+#if defined(CONFIG_SOC_SERIES_STM32C0X) || \
```
A CONFIG_FULLY_CONFIGURABLE_SEQUENCE=y/n (set depending on series) would be much more elegant

gautierg-st replied:
But for U5 it is both y and n...

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -823,6 +867,79 @@ static int set_resolution(const struct device *dev,
+	}
+
+#if defined(NOT_FULLY_CONFIGURABLE_SEQUENCER)
+#if defined(CONFIG_SOC_SERIES_STM32U5X)
+	/* On STM32U5 only ADC4 is not fully configurable */
+	if (adc == ADC4)
+#endif /* CONFIG_SOC_SERIES_STM32U5X */
```
Looks like defining this in dt would simplify code a bit

gautierg-st replied:
I could add a property for the type of sequencer, but some functions don't exist in some series, so I still need an #ifdef at some point.

gautierg-st replied:
Though I suppose I could create a macro similar to ANY_NUM_COMMON_SAMPLING_TIME_CHANNELS_IS to check if any ADC are either fully configurable or not to exclude the functions not present for the series.
I'll have a look at what is possible

gautierg-st committed:
dts: bindings: adc: stm32: add a property for adc sequencer type

Add a property for STM32 ADC to indicate which type of sequencer is used
by the device (fully configurable or not).
Add defines to help with this setting.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
dts: arm: st: add st,adc-sequencer properties to all stm32 adc

Add st,adc-sequencer to all STM32 ADC instances.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st commented:
New version adds a new property for the type of sequencer used by the ADC.
A new macro  is used to not include functions that are not defined for the series.

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -823,6 +845,72 @@ static int set_resolution(const struct device *dev,
+#endif /* ANY_ADC_SEQUENCER_TYPE_IS(FULLY_CONFIGURABLE) */
+	}
+
+#if ANY_ADC_SEQUENCER_TYPE_IS(NOT_FULLY_CONFIGURABLE)
```
Can we use `IS_ENABLED` here in order to gain readability ?
Same for all `#if ANY_ADC_SEQUENCER_TYPE_IS`

gautierg-st replied:
You mean by replacing the ANY_ADC_SEQUENCER_TYPE_IS macro entirely? I don't know how to do that with just IS_ENABLED. The problem lies with U5 that has both a fully configurable ADC (ADC1/2) and a not fully configurable one (ADC4). In this case we need to include all the code. For all other series, we include only one of the two blocks of code.

Or do you mean something like this:
```
#if ANY_ADC_SEQUENCER_TYPE_IS(NOT_FULLY_CONFIGURABLE)
#define IS_NOT_FULLY_CONFIGURABLE 1
#else
#define IS_NOT_FULLY_CONFIGURABLE 0
#endif
#if ANY_ADC_SEQUENCER_TYPE_IS(FULLY_CONFIGURABLE)
#define IS_FULLY_CONFIGURABLE 1
#else
#define IS_FULLY_CONFIGURABLE 0
#endif

#if IS_ENABLED(IS_FULLY_CONFIGURABLE)
```
That's feasible, but I don't think it makes things more readable.

erwango commented on tests/drivers/adc/adc_api/testcase.yaml:
```diff
@@ -7,3 +7,10 @@ tests:
+      DTC_OVERLAY_FILE="boards/b_u585i_iot02a_adc4.overlay"
+    platform_allow:
+      - b_u585i_iot02a
+    integration_platforms:
+      - b_u585i_iot02a
```
Is that required in this case ?

gautierg-st replied:
Probably not. I'll remove it

gautierg-st commented:
Conflict fixed

kartben commented:
Doc failure is fixed (#62967) -- you may please rebase. Sorry for the inconvenience!

gautierg-st commented:
> Doc failure is fixed (#62967) -- you may please rebase. Sorry for the inconvenience!

Rebase done. Thanks
